nuxeo-features/nuxeo-automation/nuxeo-automation-features/src/test/java/org/nuxeo/ecm/automation/core/operations/coldstorage/MoveToColdStorageTest.java

```diff
@@ -42,8 +50,44 @@
     @Inject
     protected CoreSession session;
 
+    @Inject
+    protected CoreFeature coreFeature;
+
+    @Test
+    public void shouldFailWhenUserDoNotHaveWriteColdStoragePermission() throws OperationException, IOException {
+        DocumentModel documentModel = createFileDocument(session, true);
+        ACP acp = documentModel.getACP();
+        ACL acl = acp.getOrCreateACL();
+        acl.add(new ACE("john", SecurityConstants.READ, true));
+        documentModel.setACP(acp, true);
+
+        try (CloseableCoreSession userSession = coreFeature.openCoreSession("john")) {
+            moveContentToColdStorage(userSession, documentModel);
+            fail("Should fail because the user does not have permissions to move document to cold storage");
+        } catch (NuxeoException e) {
+            assertEquals(SC_FORBIDDEN, e.getStatusCode());
+        }
+    }
+
+    @Test
+    public void shouldSucceedWhenUserHaveWriteColdStoragePermission() throws OperationException, IOException {
+        DocumentModel documentModel = createFileDocument(session, true);
+        ACP acp = documentModel.getACP();
+        ACL acl = acp.getOrCreateACL();
+        acl.add(new ACE("john", SecurityConstants.READ, true));
+        acl.add(new ACE("john", SecurityConstants.WRITE, true));
+        acl.add(new ACE("john", SecurityConstants.WRITE_COLD_STORAGE, true));
+        documentModel.setACP(acp, true);
+
+        try (CloseableCoreSession userSession = coreFeature.openCoreSession("john")) {
+            moveContentToColdStorage(userSession, documentModel);
+        } catch (NuxeoException e) {
+            fail("It was not expected an exception while moving document to cold storage since user has the required permissions");
```

Comment:
Thanks for your input!
Do you have any opinion about the "duplication" of tests here? Since we have generically the same tests on `MoveToColdStorageTest` and `TestColdStorage` classes.

The first one could, eventually, just ensure ColdStorageHelper is called and that a thumbnail is inserted as main content. But, in fact, this helper has static methods and we won't easily mock them unless we insert new dependencies and so on.
Should this helper become a service? Any clear trade off with that approach?

Reply:
I don't know the code in detail but yes test duplication is not ideal. I don't know the tests enough to give an opinion on how to best clean this up. If you decide to do a cleanup then please do it in a separate commit btw.

The helper could become a service yes, but that's more code to write. If there's a benefit then yes it could be done.

Reply:
As discussed, the main purpose of adding tests on the operation code is to ensure the correct integration between low level code (helper class) and the operation code (because the whole layer are part of cold storage feature) => mainly **Integration Test**:
- I ensure that the returned document in operation is correctly updated: the reason why I added some checks that exist on low level, plus others related to the `thumbnail`, notifications...
- on others tests on operations I followed the same logic about integration tests

I asked myself question, should I or not add the seconds check on operations, initially as i explained I tried to add a Mock like you tried, but Mockito don't allow the mock of static class, and even i tried to add the PowerMock library / dependency which allows the mock of statics class, but I abandoned it because it's more things and when we started the developments the idea to fo further in the dev, but we discussed with Florent/Antoine that perhaps in a future we will change the low level into service and i ended by adding the tests as explained above which allow more flexibility on mocking things ....

If we use this feature (cold storing) in others feature, and we test these new features than we must not do any check on the cold storage, if we do then it's a duplicate tests...


that's my point of view :)